### PR TITLE
TSPS-138 add cbas client integration

### PR DESCRIPTION
**Pull request by jsotobroad**
### Description 

This adds the ability to interact with the cbas client for a given workspace.  Currently only added the function to list all methods available.  Future work will add more functions as needed

### Jira Ticket
https://broadworkbench.atlassian.net/browse/TSPS-138

### CODE REVIEW

File: service/build.gradle [hunk 1]
```diff
@@ -35,6 +35,9 @@ dependencies {
     implementation "bio.terra:terra-common-lib" // 0.0.94-SNAPSHOT is defined in bio.terra.pipelines.java-common-conventions.gradle
     implementation "bio.terra:terra-resource-buffer-client:0.4.3-SNAPSHOT"
 
+    // hk2 is required to use cbas client, but not correctly exposed by the client
+    implementation group: "org.glassfish.jersey.inject", name: "jersey-hk2"
```
Comment (jsotobroad) [inline service/build.gradle:39]: this is pretty weird, i got this from workspace manager https://github.com/DataBiosphere/terra-workspace-manager/blob/a376aad7a021f648d7d688ba9d0555ca54ff9f1a/service/gradle/dependencies.gradle#L42 but I think we should be able to fix it on the cbas client end

Comment (mmorgantaylor) [reply service/build.gradle:39]: should we make tickets to make the fix in the cbas client and update this as necessary?

File: service/src/main/java/bio/terra/pipelines/dependencies/cbas/CbasClient.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    ApiClient apiClient = new ApiClient().setBasePath(cbasBaseUri);
+    apiClient.addDefaultHeader("Authorization", "Bearer " + accessToken);
+    // By closing the connection after each request, we avoid the problem of the open connection
+    // being force-closed ungracefully by the Azure Relay/Listener infrastructure:
+    apiClient.addDefaultHeader("Connection", "close");
+    apiClient.setDebugging(cbasConfiguration.debugApiLogging());
+    return apiClient;
+  }
+
+  PublicApi publicApi(String cbasBaseUri, String accessToken) {
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/dependencies/cbas/CbasClient.java:30]: it'd be helpful to have a brief comment for each of these describing what functionality each is used for

File: service/src/main/java/bio/terra/pipelines/service/ImputationService.java [hunk 3]
```diff
@@ -61,13 +68,29 @@ public List<ListAppResponse> queryForWorkspaceApps() {
               samService.getTspsServiceAccountToken(),
               imputationConfiguration.workspaceId()));
 
+      // cbas related calls
```
Comment (mmorgantaylor) [inline service/src/main/java/bio/terra/pipelines/service/ImputationService.java:71]: just to remind myself, this whole method `queryForWorkspaceApps()` is just a placeholder to make these calls and show they're working, right? eventually these will be moved into flight steps and the uris will be used to make calls etc etc

File: service/src/test/java/bio/terra/pipelines/configuration/external/CbasConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import bio.terra.pipelines.testutils.BaseContainerTest;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+
+class CbasConfigurationTest extends BaseContainerTest {
+  /** test reading Cbas config from application yml */
+  @Autowired CbasConfiguration cbasConfiguration;
+
+  @Test
+  void verifySamServerConfig() {
```
Comment (mmorgantaylor) [inline service/src/test/java/bio/terra/pipelines/configuration/external/CbasConfigurationTest.java:15]: minor: update copy pasta (Sam -> Cbas)

File: service/src/test/java/bio/terra/pipelines/dependencies/cbas/CbasClientTest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  String cbasBaseUri = "cbasBaseUri";
+  String authToken = "authToken";
+
+  @Test
+  void TestWdsClientApis() {
+
+    PublicApi publicApi = cbasClient.publicApi(cbasBaseUri, authToken);
+
+    assertEquals(cbasBaseUri, publicApi.getApiClient().getBasePath());
+    assertTrue(publicApi.getApiClient().isDebugging());
```
Comment (mmorgantaylor) [inline service/src/test/java/bio/terra/pipelines/dependencies/cbas/CbasClientTest.java:25]: what sets this to True? is that the default that the CBAS client sets? I guess what I'm really asking is, why are we testing this?

File: service/src/test/java/bio/terra/pipelines/dependencies/cbas/CbasServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    MethodListResponse expectedResponse =
+        new MethodListResponse().addMethodsItem(new MethodDetails().name("methodName"));
+
+    CbasClient cbasClient = mock(CbasClient.class);
+    MethodsApi methodsApi = mock(MethodsApi.class);
+    when(methodsApi.getMethods(any(), any(), any()))
+        .thenAnswer(errorAnswer)
+        .thenAnswer(errorAnswer)
+        .thenAnswer(errorAnswer)
+        .thenReturn(expectedResponse);
```
Comment (mmorgantaylor) [inline service/src/test/java/bio/terra/pipelines/dependencies/cbas/CbasServiceTest.java:79]: minor: maybe add a comment somewhere that the max number of attempts is defined as 3 in `RetryConfiguration` (this is the kind of thing that now that i've looked this up i'll never have this question again, but coming to this fresh it took me a few minutes to track that down. so not a huge deal, just would be helpful.)

**Review by mmorgantaylor — APPROVED** [review]
really nice and clear, thanks!

File: service/build.gradle (hunk 1 above)
Comment (jsotobroad) [reply service/build.gradle:39]: yup - https://broadworkbench.atlassian.net/browse/TSPS-147

File: service/src/main/java/bio/terra/pipelines/service/ImputationService.java (hunk 3 above)
Comment (jsotobroad) [reply service/src/main/java/bio/terra/pipelines/service/ImputationService.java:71]: yup yup exactly

File: service/src/test/java/bio/terra/pipelines/dependencies/cbas/CbasClientTest.java (hunk 5 above)
Comment (jsotobroad) [reply service/src/test/java/bio/terra/pipelines/dependencies/cbas/CbasClientTest.java:25]: its set to true in the application-test.yml file

File: service/src/test/java/bio/terra/pipelines/dependencies/cbas/CbasServiceTest.java (hunk 6 above)
Comment (jsotobroad) [reply service/src/test/java/bio/terra/pipelines/dependencies/cbas/CbasServiceTest.java:80]: added!

File: service/src/test/java/bio/terra/pipelines/configuration/external/CbasConfigurationTest.java (hunk 4 above)
Comment (jsotobroad) [reply service/src/test/java/bio/terra/pipelines/configuration/external/CbasConfigurationTest.java]: whoops

**Pull request merged**